src/mem/ruby/structures/RubyPrefetcherProxy.cc

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2023 ARM Limited
+ * All rights reserved
+ *
+ * The license below extends only to copyright in the software and shall
+ * not be construed as granting a license to any other intellectual
+ * property including but not limited to intellectual property relating
+ * to a hardware implementation of the functionality of the software
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mem/ruby/structures/RubyPrefetcherProxy.hh"
+
+#include "debug/HWPrefetch.hh"
+#include "mem/ruby/system/RubySystem.hh"
+
+namespace gem5
+{
+
+namespace ruby
+{
+
+RubyPrefetcherProxy::RubyPrefetcherProxy(AbstractController* _parent,
+                          prefetch::Base* _prefetcher,
+                          MessageBuffer *_pf_queue)
+    :Named(_parent->name()),
+    prefetcher(_prefetcher),
+    cacheCntrl(_parent),
+    pfQueue(_pf_queue),
+    pfEvent([this]{ issuePrefetch(); }, name()),
+    ppHit(nullptr), ppMiss(nullptr),
+    ppFill(nullptr), ppDataUpdate(nullptr)
+{
+    if (prefetcher) {
```

Comment:
Why would this ever be null? Maybe we should have an assert?

Similar for `cacheCntrl` and `pfQueue`, we should `fatal_if` these are null.

Reply:
The idea is that the proxy may not be attached to any prefetcher. This make thinks a bit cleaner on the SLICC and config side to support a config both with or without prefetcher (see the follow-up patch that adds the CHI support). The alternative would be to either conditionally create the proxy in SLICC, or let it be another simobj defined by the python script.

cacheCntrl and pfQueue I'd agree should always be non-null and will add the check.